### ISSUE TITLE
Docker: Add option to allow containers sharing a group to get unique IPs (like docker-compose)

### ISSUE THREAD

**Issue by phreakocious** [hashicorp/nomad#12815]
### Proposal
Add option (maybe in the group network stanza) to require that task containers in the same group get unique IPs.  This would reflect the behavior of docker-compose.  When listening on specific or well-known ports is necessary, it's not possible to share an IP.

### Use-cases
 - Architectures with container networks which are routable
 - Applications where hard coded or well-known ports are required

### Attempted Solutions
Placing tasks in multiple groups will get you different IPs, but obviously you lose all of the other benefits of groups.

**Comment by tgross** [comment]
Hi @phreakocious! Originally this is exactly how Nomad worked, but in 0.12.0 we deprecated task-level networking in favor of allocation-level networking. Because not all task drivers have the same networking API, we spawn a pause container (as its own "hidden" task) in order to "hold" the network namespace for other tasks. This is how we wire up bridge networking to tasks.

For the use cases you've described here:

> * Architectures with container networks which are routable

Typically this is going to be handled with a CNI plugin, but if there's a case where that doesn't work we'd like to hear about it.

> * Applications where hard coded or well-known ports are required

Typically this gets fixed via port mapping or by using `static` port allocations. There's a few different ways to work around legacy applications depending on their needs, but if you have a specific example we could probably give some suggestions for you.

I'll keep this open with a needs-discussion tag just in case some new direction can come out of it.

**Comment by phreakocious** [comment]
> Hi @phreakocious! Originally this is exactly how Nomad worked, but in 0.12.0 we deprecated task-level networking in favor of allocation-level networking. Because not all task drivers have the same networking API, we spawn a pause container (as its own "hidden" task) in order to "hold" the network namespace for other tasks. This is how we wire up bridge networking to tasks.
> 
> For the use cases you've described here:
> 
> > * Architectures with container networks which are routable
> 
> Typically this is going to be handled with a CNI plugin, but if there's a case where that doesn't work we'd like to hear about it.
> 
> > * Applications where hard coded or well-known ports are required
> 
> Typically this gets fixed via port mapping or by using `static` port allocations. There's a few different ways to work around legacy applications depending on their needs, but if you have a specific example we could probably give some suggestions for you.
> 
> I'll keep this open with a needs-discussion tag just in case some new direction can come out of it.

Yep, I'm aware of (and using) all of these since 2020.  It would be nice to avoid needing workarounds at all.  I'd like to spawn containers on a CNI macvlan network with host-local IPAM and have it look like a network of normal machines.  There is no need to use any port mappings or involve IPs on the host at all.  And also no need for a complex CNI setup with an overlay network.
